parser.mly

```diff
@@ -27,23 +27,60 @@ let scope_annotation (mode, xs) =
 program: optional_newlines prog=program_code EOF { prog }
 
 program_code:
-| prog=list(instruction_line)
+| l1=instruction_line prog=list(instruction_line) fs=list(afunction)
   {
     let annotations, instructions = List.split prog in
-    [("main",
-      (Array.of_list instructions,
-       Array.of_list annotations))]
+    let a1, i1 = l1 in
+    { main = {
+          name = "main";
+          formals = [];
+          body = [("anon", Array.of_list (i1::instructions))];
+          annotations = [("anon", Array.of_list (a1::annotations))]; };
+      functions = fs;
+    }
+  }
+| f1=afunction fs=list(afunction)
+  {
+    let fs = f1 :: fs in
+    let main = (try List.find (fun {name = n} -> n = "main") fs with
+        | Not_found -> (Printf.printf ("missing main function\n"); exit 1)
+      ) in
+    let rest = List.filter (fun {name = n} -> n <> "main") fs in
+    { main = main;
+      functions = rest; }
+  }
+
+formal_param:
+| CONST x=variable
+    { ParamConst x }
+| MUT x=variable
+    { ParamMut x }
+
+afunction:
+| FUNCTION name=variable LPAREN formals=separated_list(COMMA, formal_param) RPAREN NEWLINE optional_newlines prog=list(instruction_line)
+  {
+    let annotations, instructions = List.split prog in
```

Comment:
I addressed this by making the version a struct and moving the annotations there. That seems a lot better.
I would really not want to go back to the situation we had before where the scope module uses a different type for the function than the rest of the system. It seems that's what you are suggesting. It's what we had before and it was a big pain at the boundaries.